### ISSUE TITLE
unable to create PV in kube 1.4

### ISSUE THREAD

**Issue by Rastusik** [humblec/glusterfs-provisioner#1]
Hi, I'm using Kubernetes 1.4.4 with the newly announced StorageClass. I was able to setup a glusterfs cluster and heketi, both of which work fine outside kubernetes.

I'm having a problem while creating a PVC in kubernetes, the PVC is created correctly, the volume is created correctly using heketi, but there is a problem while creating PV for the PVC.

Somehow, this line returns an error: https://github.com/humblec/glusterfs-provisioner/blob/master/controller.go#L337

I'm geting this info from the logs:

```
I1027 09:15:21.458454       1 controller_base.go:554] storeObjectUpdate: adding claim "default/adminer-pvc", version 19068547
I1027 09:15:21.458570       1 controller.go:199] synchronizing PersistentVolumeClaim[default/adminer-pvc]: phase: Pending, bound to: "", bindCompleted: false, boundByController: false
I1027 09:15:21.458645       1 controller.go:223] synchronizing unbound PersistentVolumeClaim[default/adminer-pvc]: no volume found
I1027 09:15:21.458690       1 controller.go:1189] provisionClaim[default/adminer-pvc]: started
I1027 09:15:21.458736       1 controller.go:1353] scheduleOperation[provision-default/adminer-pvc[e2cf56c1-9c25-11e6-9ca2-fa163e5ec993]]
I1027 09:15:21.458914       1 controller.go:1208] provisionClaimOperation [default/adminer-pvc] started, class: "normal"
I1027 09:15:21.469746       1 glusterfs.go:417] glusterfs: Provison VolumeOptions {{{5368709120 0} {<nil>} 5Gi BinarySI} [ReadOnlyMany] Delete pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993 adminer-pvc kubernetes 0xc82008e728 map[resturl:http://10.254.0.15:31200] nil}
I1027 09:15:26.486920       1 controller.go:1274] volume "" for claim "default/adminer-pvc" created
I1027 09:15:26.487012       1 controller.go:1294] provisionClaimOperation [default/adminer-pvc]: trying to save volume pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993
I1027 09:15:26.487300       1 request.go:558] Request Body: "k8s\x00\n\x16\n\x02v1\x12\x10PersistentVolume\x12\xa4\x03\n\xd6\x01\n(pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993\x12\x00\x1a\x00\"\x00*\x002\x008\x00B\x00b+\n$pv.kubernetes.io/bound-by-controller\x12\x03yesb:\n\x1fpv.kubernetes.io/provisioned-by\x12\x17kubernetes.io/glusterfsb1\n'volume.beta.kubernetes.io/storage-class\x12\x06normalz\x00\x12\xbb\x01\n\x10\n\astorage\x12\x05\n\x035Gi\x12,\"*\n\x00\x12$vol_356276425a340b0548fd9c2316d3309e\x18\x00\x1a\fReadOnlyMany\"c\n\x15PersistentVolumeClaim\x12\adefault\x1a\vadminer-pvc\"$e2cf56c1-9c25-11e6-9ca2-fa163e5ec993*\x02v12\b19068547:\x00*\x06Delete\x1a\v\n\x05Bound\x12\x00\x1a\x00\x1a\x00\"\x00"
I1027 09:15:26.494571       1 controller.go:1308] failed to save volume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" for claim "default/adminer-pvc": PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value
I1027 09:15:28.619470       1 controller_base.go:582] storeObjectUpdate updating claim "default/adminer-pvc" with version 19068547
I1027 09:15:28.619567       1 controller.go:199] synchronizing PersistentVolumeClaim[default/adminer-pvc]: phase: Pending, bound to: "", bindCompleted: false, boundByController: false
I1027 09:15:28.619662       1 controller.go:223] synchronizing unbound PersistentVolumeClaim[default/adminer-pvc]: no volume found
I1027 09:15:28.619703       1 controller.go:1189] provisionClaim[default/adminer-pvc]: started
I1027 09:15:28.619744       1 controller.go:1353] scheduleOperation[provision-default/adminer-pvc[e2cf56c1-9c25-11e6-9ca2-fa163e5ec993]]
I1027 09:15:28.619782       1 controller.go:1363] operation "provision-default/adminer-pvc[e2cf56c1-9c25-11e6-9ca2-fa163e5ec993]" is already running, skipping
I1027 09:15:36.494889       1 controller.go:1294] provisionClaimOperation [default/adminer-pvc]: trying to save volume pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993
I1027 09:15:36.495483       1 request.go:558] Request Body: "k8s\x00\n\x16\n\x02v1\x12\x10PersistentVolume\x12\xa4\x03\n\xd6\x01\n(pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993\x12\x00\x1a\x00\"\x00*\x002\x008\x00B\x00b:\n\x1fpv.kubernetes.io/provisioned-by\x12\x17kubernetes.io/glusterfsb1\n'volume.beta.kubernetes.io/storage-class\x12\x06normalb+\n$pv.kubernetes.io/bound-by-controller\x12\x03yesz\x00\x12\xbb\x01\n\x10\n\astorage\x12\x05\n\x035Gi\x12,\"*\n\x00\x12$vol_356276425a340b0548fd9c2316d3309e\x18\x00\x1a\fReadOnlyMany\"c\n\x15PersistentVolumeClaim\x12\adefault\x1a\vadminer-pvc\"$e2cf56c1-9c25-11e6-9ca2-fa163e5ec993*\x02v12\b19068547:\x00*\x06Delete\x1a\v\n\x05Bound\x12\x00\x1a\x00\x1a\x00\"\x00"
I1027 09:15:36.498422       1 controller.go:1308] failed to save volume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" for claim "default/adminer-pvc": PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value
I1027 09:15:46.498795       1 controller.go:1294] provisionClaimOperation [default/adminer-pvc]: trying to save volume pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993
I1027 09:15:46.499435       1 request.go:558] Request Body: "k8s\x00\n\x16\n\x02v1\x12\x10PersistentVolume\x12\xa4\x03\n\xd6\x01\n(pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993\x12\x00\x1a\x00\"\x00*\x002\x008\x00B\x00b+\n$pv.kubernetes.io/bound-by-controller\x12\x03yesb:\n\x1fpv.kubernetes.io/provisioned-by\x12\x17kubernetes.io/glusterfsb1\n'volume.beta.kubernetes.io/storage-class\x12\x06normalz\x00\x12\xbb\x01\n\x10\n\astorage\x12\x05\n\x035Gi\x12,\"*\n\x00\x12$vol_356276425a340b0548fd9c2316d3309e\x18\x00\x1a\fReadOnlyMany\"c\n\x15PersistentVolumeClaim\x12\adefault\x1a\vadminer-pvc\"$e2cf56c1-9c25-11e6-9ca2-fa163e5ec993*\x02v12\b19068547:\x00*\x06Delete\x1a\v\n\x05Bound\x12\x00\x1a\x00\x1a\x00\"\x00"
I1027 09:15:46.502518       1 controller.go:1308] failed to save volume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" for claim "default/adminer-pvc": PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value
I1027 09:15:56.502816       1 controller.go:1294] provisionClaimOperation [default/adminer-pvc]: trying to save volume pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993
I1027 09:15:56.503307       1 request.go:558] Request Body: "k8s\x00\n\x16\n\x02v1\x12\x10PersistentVolume\x12\xa4\x03\n\xd6\x01\n(pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993\x12\x00\x1a\x00\"\x00*\x002\x008\x00B\x00b+\n$pv.kubernetes.io/bound-by-controller\x12\x03yesb:\n\x1fpv.kubernetes.io/provisioned-by\x12\x17kubernetes.io/glusterfsb1\n'volume.beta.kubernetes.io/storage-class\x12\x06normalz\x00\x12\xbb\x01\n\x10\n\astorage\x12\x05\n\x035Gi\x12,\"*\n\x00\x12$vol_356276425a340b0548fd9c2316d3309e\x18\x00\x1a\fReadOnlyMany\"c\n\x15PersistentVolumeClaim\x12\adefault\x1a\vadminer-pvc\"$e2cf56c1-9c25-11e6-9ca2-fa163e5ec993*\x02v12\b19068547:\x00*\x06Delete\x1a\v\n\x05Bound\x12\x00\x1a\x00\x1a\x00\"\x00"
I1027 09:15:56.506123       1 controller.go:1308] failed to save volume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" for claim "default/adminer-pvc": PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value
I1027 09:15:58.619867       1 controller_base.go:582] storeObjectUpdate updating claim "default/adminer-pvc" with version 19068547
I1027 09:15:58.619942       1 controller.go:199] synchronizing PersistentVolumeClaim[default/adminer-pvc]: phase: Pending, bound to: "", bindCompleted: false, boundByController: false
I1027 09:15:58.620031       1 controller.go:223] synchronizing unbound PersistentVolumeClaim[default/adminer-pvc]: no volume found
I1027 09:15:58.620074       1 controller.go:1189] provisionClaim[default/adminer-pvc]: started
I1027 09:15:58.620105       1 controller.go:1353] scheduleOperation[provision-default/adminer-pvc[e2cf56c1-9c25-11e6-9ca2-fa163e5ec993]]
I1027 09:15:58.620147       1 controller.go:1363] operation "provision-default/adminer-pvc[e2cf56c1-9c25-11e6-9ca2-fa163e5ec993]" is already running, skipping
I1027 09:16:06.506453       1 controller.go:1294] provisionClaimOperation [default/adminer-pvc]: trying to save volume pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993
I1027 09:16:06.507042       1 request.go:558] Request Body: "k8s\x00\n\x16\n\x02v1\x12\x10PersistentVolume\x12\xa4\x03\n\xd6\x01\n(pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993\x12\x00\x1a\x00\"\x00*\x002\x008\x00B\x00b+\n$pv.kubernetes.io/bound-by-controller\x12\x03yesb:\n\x1fpv.kubernetes.io/provisioned-by\x12\x17kubernetes.io/glusterfsb1\n'volume.beta.kubernetes.io/storage-class\x12\x06normalz\x00\x12\xbb\x01\n\x10\n\astorage\x12\x05\n\x035Gi\x12,\"*\n\x00\x12$vol_356276425a340b0548fd9c2316d3309e\x18\x00\x1a\fReadOnlyMany\"c\n\x15PersistentVolumeClaim\x12\adefault\x1a\vadminer-pvc\"$e2cf56c1-9c25-11e6-9ca2-fa163e5ec993*\x02v12\b19068547:\x00*\x06Delete\x1a\v\n\x05Bound\x12\x00\x1a\x00\x1a\x00\"\x00"
I1027 09:16:06.509863       1 controller.go:1308] failed to save volume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" for claim "default/adminer-pvc": PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value
I1027 09:16:16.510213       1 controller.go:1318] Error creating provisioned PV object for claim default/adminer-pvc: PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value. Deleting the volume.
I1027 09:16:16.511384       1 request.go:558] Request Body: "k8s\x00\n\v\n\x02v1\x12\x05Event\x12\xc5\x03\n5\n\x1cadminer-pvc.148158c9cd4e5341\x12\x00\x1a\adefault\"\x00*\x002\x008\x00B\x00z\x00\x12c\n\x15PersistentVolumeClaim\x12\adefault\x1a\vadminer-pvc\"$e2cf56c1-9c25-11e6-9ca2-fa163e5ec993*\x02v12\b19068547:\x00\x1a\x12ProvisioningFailed\"\xca\x01Error creating provisioned PV object for claim default/adminer-pvc: PersistentVolume \"pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993\" is invalid: spec.glusterfs.endpoints: Required value. Deleting the volume.*\x1f\n\x1bpersistentvolume-controller\x12\x002\f\b\xe0\x8b\xc7\xc0\x05\x10\xc1\xa6\xae\xf3\x01:\f\b\xe0\x8b\xc7\xc0\x05\x10\xc1\xa6\xae\xf3\x01@\x01J\aWarning\x1a\x00\"\x00"
00000060  6e 74 73 2f 61 64 6d 69  6e 65 72 2d 70 76 63 2e  |nts/adminer-pvc.|
000000e0  1a 0b 61 64 6d 69 6e 65  72 2d 70 76 63 22 24 65  |..adminer-pvc"$e|
00000170  2f 61 64 6d 69 6e 65 72  2d 70 76 63 3a 20 50 65  |/adminer-pvc: Pe|
I1027 09:16:24.525333       1 controller.go:1324] provisionClaimOperation [default/adminer-pvc]: cleaning volume pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993 succeeded
I1027 09:16:28.620208       1 controller_base.go:582] storeObjectUpdate updating claim "default/adminer-pvc" with version 19068547
I1027 09:16:28.620322       1 controller.go:199] synchronizing PersistentVolumeClaim[default/adminer-pvc]: phase: Pending, bound to: "", bindCompleted: false, boundByController: false
I1027 09:16:28.620398       1 controller.go:223] synchronizing unbound PersistentVolumeClaim[default/adminer-pvc]: no volume found
I1027 09:16:28.620434       1 controller.go:1189] provisionClaim[default/adminer-pvc]: started
I1027 09:16:28.620469       1 controller.go:1353] scheduleOperation[provision-default/adminer-pvc[e2cf56c1-9c25-11e6-9ca2-fa163e5ec993]]
I1027 09:16:28.620591       1 controller.go:1208] provisionClaimOperation [default/adminer-pvc] started, class: "normal"
```

Basically it stops after this error:

```
 controller.go:1308] failed to save volume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" for claim "default/adminer-pvc": PersistentVolume "pvc-e2cf56c1-9c25-11e6-9ca2-fa163e5ec993" is invalid: spec.glusterfs.endpoints: Required value
```

after which the gluster volume is deleted by kubernetes through heketi ... and then the process begins again and I'm stuck in the loop

Do I have to setup the glusterfs endpoints manually or is this a problem in the code?

**Comment by Rastusik** [comment]
oh nevermind, I just found this: https://github.com/kubernetes/kubernetes/pull/31854/files

but I don't understand why I can't see the info here https://github.com/kubernetes/kubernetes/tree/master/examples/experimental/persistent-volume-provisioning when the pull request was merged ...
